### PR TITLE
lua 5.4.1

### PR DESCRIPTION
**Pull request by chenrui333**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by gromgit** [comment]
Shouldn't the `test` block do an `assert_match` instead of just printing a test string?

**Comment by chenrui333** [comment]
> Shouldn't the `test` block do an `assert_match` instead of just printing a test string?

I was literally thinking the same thing just now :D

I will update that.

**Comment by chenrui333** [comment]
Should be good now.

File: Formula/lua.rb [hunk 1]
```diff
@@ -80,68 +79,68 @@ def caveats
...
 end
 
 __END__
 diff --git a/Makefile b/Makefile
-index 7fa91c8..a825198 100644
+index 416f444..1424dba 100644
 --- a/Makefile
 +++ b/Makefile
-@@ -41,7 +41,7 @@ PLATS= aix bsd c89 freebsd generic linux macosx mingw posix solaris
+@@ -41,7 +41,7 @@ PLATS= guess aix bsd c89 freebsd generic linux linux-readline macosx mingw posix
```
Comment (miccal) [inline Formula/lua.rb:91]: I think you have mistakenly removed `solaris` here.

Comment (chenrui333) [reply Formula/lua.rb:91]: This is actually generated, not done by hand.

Comment (chenrui333) [reply Formula/lua.rb:91]: I dont know why it dropped the `solaris` in the end, but it would apply into the diff though, thanks for catching this out. :)

Comment (miccal) [reply Formula/lua.rb:91]: You're welcome @chenrui333 -- I was working on this myself earlier today, but you beat me to it.

**Comment by chenrui333** [comment]
10.13

```
Error: 17 failed steps!
brew test --verbose luarocks
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle instead
brew install --verbose --build-bottle kyoto-tycoon
brew install --verbose --build-bottle lmod
brew audit --skip-style lmod --online --git
brew install --verbose --build-bottle lsyncd
brew install --verbose --build-bottle macvim
brew install --verbose --build-bottle mame
brew install --verbose --build-bottle neomutt
brew install --verbose --build-bottle osrm-backend
brew audit --skip-style pdns --online --git
brew install --verbose --build-bottle pdnsrec
brew install --verbose --build-bottle sile
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle vim
brew install --verbose --build-bottle vis
```

10.14
```
Error: 18 failed steps!
brew test --verbose luarocks
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle instead
brew install --verbose --build-bottle kyoto-tycoon
brew install --verbose --build-bottle lmod
brew audit --skip-style lmod --online --git
brew install --verbose --build-bottle lsyncd
brew install --verbose --build-bottle macvim
brew install --verbose --build-bottle mame
brew install --verbose --build-bottle neomutt
brew install --verbose --build-bottle osrm-backend
brew audit --skip-style pdns --online --git
brew install --verbose --build-bottle pdnsrec
brew install --verbose --build-bottle sile
brew install --verbose --build-bottle tracebox
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle vim
brew install --verbose --build-bottle vis
##[error]Process completed with exit code 1.
```

10.15

```
Error: 19 failed steps!
brew test --verbose luarocks
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle instead
brew install --verbose --build-bottle kyoto-tycoon
brew install --verbose --build-bottle lmod
brew audit --skip-style lmod --online --git
brew fetch --retry lsyncd --build-bottle --force
brew install --verbose --build-bottle lsyncd
brew install --verbose --build-bottle macvim
brew install --verbose --build-bottle mame
brew install --verbose --build-bottle neomutt
brew install --verbose --build-bottle osrm-backend
brew audit --skip-style pdns --online --git
brew install --verbose --build-bottle pdnsrec
brew install --verbose --build-bottle sile
brew install --verbose --build-bottle tracebox
brew audit --skip-style tracebox --online --git
brew install --verbose --build-bottle vim
brew install --verbose --build-bottle vis
```

**Comment by chenrui333** [comment]
rebased to latest master and fixed the merge conflict with vim

File: Formula/lua.rb (hunk 1 above)
Comment (alebcay) [reply Formula/lua.rb:91]: Looks like `solaris` is still missing here - does it need to be added back?

Comment (alebcay) [inline Formula/lua.rb:91]: Looks like `solaris` is still missing here - does it need to be added back?

**Comment by alerque** [comment]
Lua 5.4 support for SILE is ready to go upstream ([tracking issue](https://github.com/sile-typesetter/sile/issues/668), [pending PR](https://github.com/sile-typesetter/sile/pull/929)), but it's held up being released until there is a published version of lua-zlib that works ([tracking issue](https://github.com/brimworks/lua-zlib/issues/49)). I'll be sure to cut a release as soon as that's available, at which point SILE won't be a hold up for Homebrew switching to Lua 5.4!

**Comment by alerque** [comment]
As of v0.10.9 SILE supports Lua 5.4 ([Homebrew PR here](https://github.com/Homebrew/homebrew-core/pull/58199)). After that is merged the formula can be updated to use Lua 5.4 at any time. Note some of the LuaRocks being installed are being built from the Github tagged versions. Several projects have released new Rocks on LuaRocks by bumping the rockrel without tagging a release on Github. For every project where that is the case the upper bound restriction can just be removed from the rockspec before building. If you want me to submit a patch with `sed command to take care of that universally for that formula I can.

File: Formula/lua.rb (hunk 1 above)
Comment (chenrui333) [reply Formula/lua.rb:91]: This won't impact the git patch. (more just like reference line)

**Comment by chenrui333** [comment]
Going to refresh this PR now.

**Comment by chenrui333** [comment]
side note here, will try to add lua@5.3 for the failed formulae builds

**Comment by nloveladyallen** [comment]
Fixing luarocks is easy:

```diff
--- a/Formula/luarocks.rb
+++ b/Formula/luarocks.rb
@@ -35,17 +35,17 @@ class Luarocks < Formula
   end

   test do
-    ENV["LUA_PATH"] = "#{testpath}/share/lua/5.3/?.lua"
-    ENV["LUA_CPATH"] = "#{testpath}/lib/lua/5.3/?.so"
+    ENV["LUA_PATH"] = "#{testpath}/share/lua/5.4/?.lua"
+    ENV["LUA_CPATH"] = "#{testpath}/lib/lua/5.4/?.so"

-    (testpath/"lfs_53test.lua").write <<~EOS
+    (testpath/"lfs_54test.lua").write <<~EOS
       require("lfs")
       print(lfs.currentdir())
     EOS

     system "#{bin}/luarocks", "--tree=#{testpath}", "install", "luafilesystem"
     system "lua", "-e", "require('lfs')"
-    assert_match testpath.to_s, shell_output("lua lfs_53test.lua")
+    assert_match testpath.to_s, shell_output("lua lfs_54test.lua")

     ENV["LUA_PATH"] = "#{testpath}/share/lua/5.1/?.lua"
     ENV["LUA_CPATH"] = "#{testpath}/lib/lua/5.1/?.so"
```

There will still have to be a lua@5.3 formula because Lua 5.4 isn't backwards compatible (take a look at the build errors for dependents—a lot are API changes, not just build systems expecting 5.3). I'd take a shot at it myself but I don't have a macOS VM set up and I don't want to screw up my local Lua install 😃

**Comment by chenrui333** [comment]
Just made a version bump for lurarocks today, #61916

**Comment by ByunghoonKim** [comment]
Please note that Lua 5.4.1 and Lua 5.3.6 have been released. [Lua: news](https://www.lua.org/news.html)

**Comment by fxcoudert** [comment]
Updated to 5.4.1 and rebased

**Comment by fxcoudert** [comment]
```
brew test --retry --verbose aspcud
brew linkage --test clingo
brew test --retry --verbose clingo
brew linkage --test freeswitch
brew linkage --test lcm
brew test --verbose luarocks
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle instead
brew install --verbose --build-bottle lmod
brew fetch --retry lsyncd --build-bottle --force
brew install --verbose --build-bottle 
brew install --verbose --build-bottle osrm-
brew install --verbose --build-bottle pdnsrec
brew install --verbose --build-bottle sile
brew install --verbose --build-bottle tracebox
brew audit tracebox --online --git --skip-style
brew install --verbose --build-bottle vis
```

**Comment by fxcoudert** [comment]
Help welcome from anyone who knows how lua/luarocks work, because I don't 😢

**Comment by fxcoudert** [comment]
- lsyncd https://github.com/axkibe/lsyncd/issues/621
- corsixth https://github.com/CorsixTH/CorsixTH/pull/1686
- instead https://github.com/instead-hub/instead/commit/ea4f0e81c6859b0aadde582f47ffd7850f54a264

**Comment by SMillerDev** [comment]
10.14/10.15
```
Error: 13 failed steps!
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle kyoto-tycoon
brew install --verbose --build-bottle lmod
brew fetch --retry lsyncd --build-bottle --force
brew install --verbose --build-bottle lsyncd
brew audit macvim --online --git --skip-style
brew audit neomutt --online --git --skip-style
brew audit ntopng --online --git --skip-style
brew install --verbose --build-bottle osrm-backend
brew install --verbose --build-bottle sile
brew install --verbose --build-bottle tracebox
brew audit tracebox --online --git --skip-style
brew install --verbose --build-bottle vis
```

10.13
```
Error: 12 failed steps!
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle kyoto-tycoon
brew install --verbose --build-bottle lmod
brew install --verbose --build-bottle lsyncd
brew audit neomutt --online --git --skip-style
brew audit ntopng --online --git --skip-style
brew audit osm2pgsql --online --git --skip-style
brew install --verbose --build-bottle osrm-backend
brew audit osrm-backend --online --git --skip-style
brew install --verbose --build-bottle sile
brew audit tracebox --online --git --skip-style
brew install --verbose --build-bottle vis
```

**Comment by tobylane** [comment]
Corsixth 0.64 needs Lua 5.1-3, for you lua@5.3. 0.65 will work with 5.4.

**Comment by chenrui333** [comment]
going to pick up tonight.

**Comment by chenrui333** [comment]
```
Error: 12 failed steps!
brew test --retry --verbose aspcud
brew test --retry --verbose clingo
brew test --verbose luarocks
brew install --verbose --build-bottle corsixth
brew install --verbose --build-bottle instead
brew install --verbose --build-bottle lmod
brew install --verbose --build-bottle lsyncd
brew install --verbose --build-bottle osrm-backend
brew install --verbose --build-bottle sile
brew install --verbose --build-bottle tracebox
brew audit tracebox --online --git --skip-style
brew install --verbose --build-bottle vis
```

**Comment by carlocab** [comment]
I tried opening a PR to try to update lua, but I can now see that that's not going to work.

Would the following work?
1. Add a `lua@5.3` formula in a separate PR
2. When the previous is merged, any formulae causing test failures here can have their dependency changed to `lua@5.3` so this PR can be merged.
3. Try to bump the formulae that had to depend on `lua@5.3` to `lua@5.4` in separate PRs.

**Comment by fxcoudert** [comment]
@carlocab all formulas that are not currently working with lug 5.4 need to be investigated: is upstream aware of the issue, when is a fix planned, etc. We don't want to migrate formulas to an older version if they are not going to be migrated over time to something more recent

**Comment by carlocab** [comment]
Thanks @fxcoudert. I'm aware of that. That's what step 3 in my last comment was for.

I'm just thinking it makes more sense to migrate all the formulae that get stuck on `lua@5.3` slowly individually, rather than have them block like upgrade to lua like they're doing in this PR.

I think my suggestion to migrate formulae separately rather all in one big PR also benefits from allowing the community to pitch in more with the migration.

I would also like to point out that different versions of lua aren't quite like different versions of other packages. Here's Neovim's lead maintainer on the subject:

https://github.com/neovim/neovim/issues/9760#issuecomment-474340616

**Comment by alerque** [comment]
@fxcoudert I hear what you are saying, but I don't think it quite fits the circumstances here. What @carlocab proposed isn't "migrating formulas to and older version" *per se*, it is holding them on the current version they are using successfully while not holding back other projects that could/should use a new Lua release. Unfortunately the Lua ecosystem is somewhat fragmented and different language VMs have different features. 5.1→5.2→5.3 is not as straightforward a version bump upgrade path as the numbers would lead you to believe. Some projects will require 5.3 for some time yet. Some may never get ported, but that doesn't make them obsolete. There are actively maintained projects out there that still depend on 5.1 language features. This shouldn't by such a road block to updating the default Lua interpreter.

The proposal here to upgrade everything that can be while not needing to hold forever or mass-remove a bunch of working formula makes great sense given the state of the Lua ecosystem. I don't *like* it either, but I don't manage Lua releases upstream :man_shrugging:. As a downstream Lua developer quite familiar with different projects' approaches to interpreter versioning I can vouch for the sanity of this idea.

**Comment by gromgit** [comment]
From https://www.lua.org/versions.html#numbering:
> Different versions are really different. The API is likely to be a little different (but with compatibility switches), and there is no ABI compatibility: applications that embed Lua and C libraries for Lua must be recompiled. The virtual machine is also very likely to be different in a new version: Lua programs that have been precompiled for one version will not load in a different version.

The way I read it, applications stick with a specific Lua version to embed, and move only when required. Also, the "top-level" Lua dylib symlinks actually embed the version in their names (`liblua5.1.dylib` vs. `liblua5.3.dylib`), just like on every Linux distro I've seen (`liblua5.1.so`, etc.). Also, AFAIK, only Homebrew specifically creates a `liblua.dylib` symlink, contrary to every other installation I've seen.

IMO, Lua versions should be treated as completely different entities, not different versions of the same language, so a saner naming scheme that causes the fewest problems in the long term is `lua51`, `lua53`, etc. Since pretty much nothing depends on Lua 5.4 at this point, can I suggest naming this formula `lua54` instead, and leaving the other `lua` formulae alone for now?

**Comment by fxcoudert** [comment]
> can I suggest naming this formula lua54 instead, and leaving the other lua formulae alone for now?

No, please don't.

We are aware that lua has several versions with differences, and therefore a `lua@5.3` will be accepted in homebrew-core, as that is still maintained. However, we also need to make sure that there is some movement of the other formulas depending on it, otherwise we end up with unmaintainable messes: for example, we still have formulas that depend on lua@5.1 even though that has been unmaintained for 8 years (last release in Feb 2012);

**Comment by fxcoudert** [comment]
> I'm aware of that. That's what step 3 in my last comment was for. (We can even open a tracking issue for it...)

@carlocab excellent then. Only thing is: this cannot be done in separate PRs, it needs to be the same PR that moves lua to 5.4, creates a lua@5.3, rebuilds the formulas that build against 5.4, and migrate the other ones to 5.3

I only wish to make sure, before we accept that, that all formulas that are "stuck" on 5.3 have been notified, and we add a link to the upstream tracking issue above the dependency in the formula. That way, we can check we when update them, if support for latest lua is available.

**Comment by carlocab** [comment]
> excellent then. Only thing is: this cannot be done in separate PRs, it needs to be the same PR that moves lua to 5.4, creates a lua@5.3, rebuilds the formulas that build against 5.4, and migrate the other ones to 5.3

I may have expressed myself poorly then because this is exactly what I meant. Apologies for the miscommunication.

To be clear, step 3 was meant to handle the formulae that get stuck with `lua@5.3` from this PR.

**Comment by fxcoudert** [comment]
Merged in https://github.com/Homebrew/homebrew-core/pull/65993

**Pull request closed without merging**